File: python/2015/day14.py
# python day14.py ../../data/2015/day14.dat

import sys

roster = {}


def get_roster(data):
    global roster
    for deer in data:
        stats = deer.split()
        name = stats[0]
        speed = int(stats[3])
        duration = int(stats[6])
        cooldown = int(stats[13])
        roster[name] = list(
            [speed for i in range(duration)] + [0 for i in range(cooldown)]
        )


def parse_input(input_data):
    """
    Parse the input data from the Advent of Code problem.
    Edit this function to fit the day's input format.
    """
    return input_data.strip().split("\n")


def part1(data):
    global roster
    roster.clear()
    get_roster(data)

    log = {deer: 0 for deer in roster}

    for t in range(2503):
        for deer in roster:
            log[deer] += roster[deer][t % len(roster[deer])]

    return max(log.values())


def part2(data):
    global roster
    roster.clear()
    get_roster(data)

    log = {deer: 0 for deer in roster}
    points = {deer: 0 for deer in roster}

    for t in range(2503):
        for deer in roster:
            log[deer] += roster[deer][t % len(roster[deer])]

        winner = max(log.values())
        awards = [k for k, v in log.items() if v == winner]
        for deer in awards:
            points[deer] += 1

    return max(points.values())


if __name__ == "__main__":
    # Read input file or use test string
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            input_data = f.read()
    else:
        # Example test input, replace with actual test data
        input_data = """\
Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.
"""

    data = parse_input(input_data)
    print("Day 14 Part 1:", part1(data))
    print("Day 14 Part 2:", part2(data))

File: java/2015/Day14.java
import module java.base;

public class Day14 {
    private record Deer(String name, int speed, int fly, int rest) {}

    public static void main(String[] args) {
        var input = readInput("../../data/2015/day14.dat");
        if (input.isEmpty()) return;

        var pattern = Pattern.compile("(\\w+) can fly (\\d+) km/s for (\\d+) seconds, but then must rest for (\\d+) seconds\\.");
        var deers = new ArrayList<Deer>();
        for (var line : input.lines().toList()) {
            var m = pattern.matcher(line);
            if (!m.matches()) continue;
            deers.add(new Deer(m.group(1),
                Integer.parseInt(m.group(2)),
                Integer.parseInt(m.group(3)),
                Integer.parseInt(m.group(4))));
        }

        var distances = new int[deers.size()];
        var points = new int[deers.size()];
        var maxDist = 0;

        for (var t = 0; t < 2503; t++) {
            maxDist = 0;
            for (var i = 0; i < deers.size(); i++) {
                var d = deers.get(i);
                if (t % (d.fly() + d.rest()) < d.fly()) {
                    distances[i] += d.speed();
                }
                maxDist = Math.max(maxDist, distances[i]);
            }
            for (var i = 0; i < deers.size(); i++) {
                if (distances[i] == maxDist) points[i]++;
            }
        }

        System.out.println("Part 1: " + Arrays.stream(distances).max().getAsInt());
        System.out.println("Part 2: " + Arrays.stream(points).max().getAsInt());
    }

    private static String readInput(String filePath) {
        try {
            return Files.readString(Paths.get(filePath)).trim();
        } catch (NoSuchFileException e) {
            System.out.println("Data file missing: " + filePath);
        } catch (IOException ex) {
        }
        return "";
    }
}
